client/src/main/scala/org/http4s/client/oauth1.scala

```diff
@@ -0,0 +1,102 @@
+      val auth = genAuthHeader(req.method, req.uri, params, consumer, callback, verifier, token)
+      req.withHeaders(auth)
+    }
+  }
+
+  def genAuthHeader(method: Method, uri: Uri, userParams: Seq[(String,String)], consumer: Consumer,
+                    callback: Option[Uri], verifier: Option[String], token: Option[Token]): Authorization = {
+    val params = {
+      val params = new ListBuffer[(String,String)]
+      params += "oauth_consumer_key"     -> encode(consumer.key)
+      params += "oauth_signature_method" -> "HMAC-SHA1"
+      params += "oauth_timestamp"        -> (System.currentTimeMillis / 1000).toString
+      params += "oauth_nonce"            -> System.nanoTime.toString
+      params += "oauth_version"          -> "1.0"
+      params += "oauth_callback"         -> callback.map(c => encode(c.renderString)).getOrElse(OutOfBounds)
+      token.foreach { t => params += "oauth_token" -> encode(t.value) }
+      verifier.foreach { v => params += "oauth_verifier" -> encode(v) }
+      params.result()
+    }
+
+    val baseString = genBaseString(method, uri, params ++ userParams.map{ case (k,v) => (encode(k), encode(v))})
+    val sig = makeSHASig(baseString, consumer, token)
+    val creds = GenericCredentials("OAuth".ci, params.toMap + ("oauth_signature" -> encode(sig)))
+
+    Authorization(creds)
+  }
+
+  // baseString must already be encoded, consumer and token must not be
+  private[client] def makeSHASig(baseString: String, consumer: Consumer, token: Option[Token]): String = {
+    val sha1 = crypto.Mac.getInstance(SHA1)
+    val key = encode(consumer.secret) + "&" + token.map(t => encode(t.secret)).getOrElse("")
+    sha1.init(new crypto.spec.SecretKeySpec(bytes(key), SHA1))
+
+    val sigBytes = sha1.doFinal(bytes(baseString))
+    net.iharder.Base64.encodeBytes(sigBytes)
+  }
+
+  // Needs to have all params already encoded
+  private[client] def genBaseString(method: Method, uri: Uri, params: Seq[(String,String)]): String = {
+    val paramsStr = params.map{ case (k,v) => k + "=" + v }.sorted.mkString("&")
+
+    Seq(method.name,
+      encode(uri.copy(query = Query.empty, fragment = None).renderString),
+      encode(paramsStr)
+    ).mkString("&")
+  }
+
+  private[client] def encode(str: String): String =
+    UrlCodingUtils.urlEncode(str, spaceIsPlus = false, toSkip = UrlFormCodec.urlUnreserved)
+
+  private[http4s] def getUserParams(req: Request): Task[(Request,Seq[(String, String)])] = {
+    val qparams = req.uri.query.map{ case (k,ov) => (k, ov.getOrElse("")) }
+
+    req.contentType match {
+      case Some(t) if (req.method == Method.POST || req.method == Method.PUT) &&
+                       t.mediaType == MediaType.`application/x-www-form-urlencoded` =>
+        req.as[UrlForm].flatMap { urlform =>
+          val bodyparams = urlform.values.toSeq
+            .flatMap{ case (k,vs) => if (vs.isEmpty) Seq(k->"") else vs.map((k,_))}
+
+          req.withBody(urlform)(UrlForm.entityEncoder(Charset.`UTF-8`))
```

Comment:
You're imposing UTF-8 when the request may have been something different.  I think it's correct, but is this rude?